crates/circuit/src/circuit_instruction.rs

```diff
@@ -51,32 +56,111 @@ impl ExtraInstructionAttributes {
         duration: Option<Py<PyAny>>,
         unit: Option<String>,
         condition: Option<Py<PyAny>>,
-    ) -> Option<Self> {
-        if label.is_some() || duration.is_some() || unit.is_some() || condition.is_some() {
-            Some(Self {
-                label,
-                duration,
-                unit,
-                condition,
-            })
-        } else {
-            None
+    ) -> Self {
+        ExtraInstructionAttributes {
+            attributes: if label.is_some()
+                || duration.is_some()
+                || unit.is_some()
+                || condition.is_some()
+            {
+                Some(Box::new(ExtraAttrs {
+                    label,
+                    duration,
+                    unit,
+                    condition,
+                }))
+            } else {
+                None
+            },
         }
     }
 
     /// Get the Python-space version of the stored `unit`.  This evalutes the Python-space default
     /// (`"dt"`) value if we're storing a `None`.
     pub fn py_unit(&self, py: Python) -> Py<PyString> {
-        self.unit
+        self.attributes
             .as_deref()
-            .map(|unit| <&str as IntoPy<Py<PyString>>>::into_py(unit, py))
+            .and_then(|attrs| {
+                attrs
+                    .unit
+                    .as_deref()
+                    .map(|unit| <&str as IntoPy<Py<PyString>>>::into_py(unit, py))
+            })
             .unwrap_or_else(|| Self::default_unit(py).clone().unbind())
     }
 
     /// Get the Python-space default value for the `unit` field.
     pub fn default_unit(py: Python) -> &Bound<PyString> {
         intern!(py, "dt")
     }
+
+    pub fn label(&self) -> Option<&str> {
+        self.attributes
+            .as_deref()
+            .and_then(|attrs| attrs.label.as_deref())
+    }
+
+    pub fn set_label(&mut self, label: Option<String>) {
+        if let Some(attrs) = &mut self.attributes {
+            attrs.label = label;
+            self.shrink_if_empty();
+            return;
+        }
+        if label.is_some() {
+            self.attributes = Some(Box::new(ExtraAttrs {
+                label,
+                duration: None,
+                unit: None,
+                condition: None,
+            }))
+        }
+    }
+
+    pub fn duration(&self) -> Option<&PyObject> {
+        self.attributes
+            .as_deref()
+            .and_then(|attrs| attrs.duration.as_ref())
+    }
+
+    pub fn unit(&self) -> Option<&str> {
+        self.attributes
+            .as_deref()
+            .and_then(|attrs| attrs.unit.as_deref())
+    }
```

Comment:
Do you want to put the setters for `duration` and `unit` on now, or wait til we need them?

Reply:
Done in bca33aa9443b0122a8411e057e40f63b7d5eb59f.